src/utils/test/challenge/eval-code-run-tests.js

```diff
@@ -1,35 +1,41 @@
 import executeTests from './execute-tests'
 import { SUPPRESS_TESTS } from '../../regexp'
 import TESTS from '../../../assets/testRef'
-import loopProtect from './loop-protect'
+import wontTimeout from './loop-protect';
+/* eslint-disable no-eval */
 
 export default (code, id) => {
-  try {
-    /* eslint-disable no-unused-vars */
-    const assert = require('assert')
-    const { forEach, invoke } = require('lodash')
-    const { beginTests, logTestReport } = require('./testReport')
-    const isTestDisabled = require('../common/is-test-disabled')
-    /* eslint-enable no-unused-vars */
 
-    let prepend = 'const tests = ', tail = '', tests = ''
+  // import funcs used in executeTests so in
+  // scope during eval, see ./execute-tests.js
 
-    // if suppressTests is true, only eval code,
-    // otherwise, eval code and run tests (if tests exist)
-    if (!SUPPRESS_TESTS.test(code) && TESTS[id]) {
-      tests = prepend + JSON.stringify(TESTS[id].tests)
-      if (TESTS[id].tail) tail += TESTS[id].tail
-    }
-    
-    // eslint-disable-next-line
-    eval(
-      loopProtect(code) +
```

Comment:
for some reason this caused problems with loop-free code. Tests would multiply (like 700 of them) and they would all log as `Fail: undefined`. I couldn't really see why this was happening, but figured out that eval'ing the user code on its own without the tests prevented this, so devised a way to test code first for loops, then if infinite-loop free, go on to test. Throw custom error if infinite loop. Simultaneously realized that replacement logic wouldn't work if multiple loops were present, so fixed that as well.